<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>MGDrivE2: Data Storage and Analysis</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>
<style type="text/css">
a.anchor-section {margin-left: 10px; visibility: hidden; color: inherit;}
a.anchor-section::before {content: '#';}
.hasAnchor:hover a.anchor-section {visibility: visible;}
</style>
<script>// Anchor sections v1.0 written by Atsushi Yasumoto on Oct 3rd, 2020.
document.addEventListener('DOMContentLoaded', function() {
  // Do nothing if AnchorJS is used
  if (typeof window.anchors === 'object' && anchors.hasOwnProperty('hasAnchorJSLink')) {
    return;
  }

  const h = document.querySelectorAll('h1, h2, h3, h4, h5, h6');

  // Do nothing if sections are already anchored
  if (Array.from(h).some(x => x.classList.contains('hasAnchor'))) {
    return null;
  }

  // Use section id when pandoc runs with --section-divs
  const section_id = function(x) {
    return ((x.classList.contains('section') || (x.tagName === 'SECTION'))
            ? x.id : '');
  };

  // Add anchors
  h.forEach(function(x) {
    const id = x.id || section_id(x.parentElement);
    if (id === '') {
      return null;
    }
    let anchor = document.createElement('a');
    anchor.href = '#' + id;
    anchor.classList = ['anchor-section'];
    x.classList.add('hasAnchor');
    x.appendChild(anchor);
  });
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">MGDrivE2: Data Storage and Analysis</h1>



<div id="table-of-contents" class="section level2">
<h2>Table of Contents</h2>
<ol style="list-style-type: decimal">
<li><a href="#pars">Parameterization</a></li>
<li><a href="#init_spn">Initialization of the Petri Net</a></li>
<li><a href="#equilibria_haz">Equilibrium Conditions and Hazard Functions</a></li>
<li><a href="#sim">Simulation of Fully Specified SPN Model</a>
<ol style="list-style-type: decimal">
<li><a href="#soln_tau">Tau-leaping Simulation</a></li>
<li><a href="#split_agg">Trace Analysis</a></li>
<li><a href="#summary">Summary Analysis</a></li>
</ol></li>
<li><a href="#ref">References</a></li>
</ol>
<div id="preface" class="section level3">
<h3>Preface</h3>
<p>There are several vignettes explaining simulation setup, for basic lifecycle simulations <a href="lifecycle-node.html">here</a> and <a href="lifecycle-network.html">here</a>, as well as for human/mosquito epidemiological studies, <a href="epi-node.html">here</a> and <a href="epi-network.html">here</a> for <strong>SEI</strong>/<strong>SIS</strong> models and <a href="epi-SEIR.html">here</a> for <strong>SEI</strong>/<strong>SEIR</strong> models. However, for actual explorations, many stochastic realizations are needed, as well as summary statistics of those simulations. <strong>MGDrivE2</strong> provides tools for storing output, as well as analyzing and summarizing that output. This vignette will use the <a href="epi-network.html">network <strong>SEI</strong>/<strong>SIS</strong></a> simulation as the example simulation to showcase data storage and analysis.</p>
<p>We start by loading the <strong>MGDrivE2</strong> package, as well as the <strong>MGDrivE</strong> package for access to inheritance cubes, and <strong>Matrix</strong> for sparse matrices used in migration. We will use the basic cube to simulate Mendelian inheritance for this example.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="co"># simulation functions</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a><span class="kw">library</span>(MGDrivE2)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a><span class="co"># inheritance patterns</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a><span class="kw">library</span>(MGDrivE)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true"></a><span class="co"># sparse migration</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true"></a><span class="kw">library</span>(Matrix)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true"></a><span class="co"># basic inheritance pattern</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true"></a>cube &lt;-<span class="st"> </span>MGDrivE<span class="op">::</span><span class="kw">cubeMendelian</span>()</span></code></pre></div>
</div>
</div>
<div id="pars" class="section level2">
<h2>Parameterization</h2>
<p>These are the same parameters found in the <a href="epi-network.html">epi-network</a> vignette, and are also explained in the <a href="lifecycle-node.html">lifecycle-node</a> vignette.</p>
<p>We will store output every other day (<code>dt = 2</code>), as writing to disk can be slow and this reduces the amount of data to store by half. The rate that one needs to store data will depend on the accuracy of one’s parameters - i.e., if your parameters are highly questionable, storing data every other day will not lose much information, but if your parameters are very well defined, daily storage may be necessary for accuracy. Note we set <code>tmax</code> artifically low in this vignette to reduce build times as the goal of this article is to demonstrate CSV processing capabilities of <strong>MGDrivE2</strong>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="co"># entomological and epidemiological parameters</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true"></a>theta &lt;-<span class="st"> </span><span class="kw">list</span>(</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true"></a>  <span class="co"># lifecycle parameters</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true"></a>  <span class="dt">qE =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">4</span>,</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true"></a>  <span class="dt">nE =</span> <span class="dv">2</span>,</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true"></a>  <span class="dt">qL =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">3</span>,</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true"></a>  <span class="dt">nL =</span> <span class="dv">3</span>,</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true"></a>  <span class="dt">qP =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">6</span>,</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true"></a>  <span class="dt">nP =</span> <span class="dv">2</span>,</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true"></a>  <span class="dt">muE =</span> <span class="fl">0.05</span>,</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true"></a>  <span class="dt">muL =</span> <span class="fl">0.15</span>,</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true"></a>  <span class="dt">muP =</span> <span class="fl">0.05</span>,</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true"></a>  <span class="dt">muF =</span> <span class="fl">0.09</span>,</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true"></a>  <span class="dt">muM =</span> <span class="fl">0.09</span>,</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true"></a>  <span class="dt">beta =</span> <span class="dv">16</span>,</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true"></a>  <span class="dt">nu =</span> <span class="dv">1</span><span class="op">/</span>(<span class="dv">4</span><span class="op">/</span><span class="dv">24</span>),</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true"></a>  <span class="co"># epidemiological parameters</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true"></a>  <span class="dt">NH =</span> <span class="dv">1000</span>,</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true"></a>  <span class="dt">X =</span> <span class="fl">0.25</span>,</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true"></a>  <span class="dt">f =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">3</span>,</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true"></a>  <span class="dt">Q =</span> <span class="fl">0.9</span>,</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true"></a>  <span class="dt">b =</span> <span class="fl">0.55</span>,</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true"></a>  <span class="dt">c =</span> <span class="fl">0.15</span>,</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true"></a>  <span class="dt">r =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">200</span>,</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true"></a>  <span class="dt">muH =</span> <span class="dv">1</span><span class="op">/</span>(<span class="dv">62</span><span class="op">*</span><span class="dv">365</span>),</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true"></a>  <span class="dt">qEIP =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">11</span>,</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true"></a>  <span class="dt">nEIP =</span> <span class="dv">2</span></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true"></a>)</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true"></a></span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true"></a><span class="co"># simulation parameters</span></span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true"></a>tmax &lt;-<span class="st"> </span><span class="dv">50</span></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true"></a>dt &lt;-<span class="st"> </span><span class="dv">2</span></span></code></pre></div>
<p>Additionally, we augment the cube with transmission efficiencies; see Smith &amp; McKenzie (2004) for an extensive discussion of these parameters.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a><span class="co"># augment the cube with RM transmission parameters</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a>cube<span class="op">$</span>c &lt;-<span class="st"> </span><span class="kw">setNames</span>(<span class="dt">object =</span> <span class="kw">rep</span>(<span class="dt">x =</span> theta<span class="op">$</span>c, <span class="dt">times =</span> cube<span class="op">$</span>genotypesN), <span class="dt">nm =</span> cube<span class="op">$</span>genotypesID)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a>cube<span class="op">$</span>b &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;AA&quot;</span> =<span class="st"> </span>theta<span class="op">$</span>b, <span class="st">&quot;Aa&quot;</span> =<span class="st"> </span><span class="fl">0.35</span>, <span class="st">&quot;aa&quot;</span> =<span class="st"> </span><span class="dv">0</span>)</span></code></pre></div>
</div>
<div id="init_spn" class="section level2">
<h2>Initialization of the Petri Net</h2>
<p>With the basic parameters finished, we setup the types of nodes in our network, their movement in relation to each other, and finally the structure of the Petri Net.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a><span class="co"># nodetypes</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true"></a>node_list &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;m&quot;</span>, <span class="st">&quot;b&quot;</span>, <span class="st">&quot;h&quot;</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true"></a>num_nodes &lt;-<span class="st"> </span><span class="kw">length</span>(node_list)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true"></a><span class="co"># human movement</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true"></a>h_move &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dt">data =</span> <span class="ot">FALSE</span>, <span class="dt">nrow =</span> num_nodes, <span class="dt">ncol =</span> num_nodes,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true"></a>                 <span class="dt">dimnames =</span> <span class="kw">list</span>(node_list, node_list))</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true"></a>h_move[<span class="dv">2</span>,<span class="dv">3</span>] &lt;-<span class="st"> </span><span class="ot">TRUE</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true"></a>h_move[<span class="dv">3</span>,<span class="dv">2</span>] &lt;-<span class="st"> </span><span class="ot">TRUE</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true"></a><span class="co"># mosquito movement</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true"></a>m_move &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dt">data =</span> <span class="ot">FALSE</span>, <span class="dt">nrow =</span> num_nodes, <span class="dt">ncol =</span> num_nodes,</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true"></a>                 <span class="dt">dimnames =</span> <span class="kw">list</span>(node_list, node_list))</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true"></a>m_move[<span class="dv">1</span>,<span class="dv">2</span>] &lt;-<span class="st"> </span><span class="ot">TRUE</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true"></a>m_move[<span class="dv">2</span>,<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="ot">TRUE</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true"></a></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true"></a><span class="co"># Places and transitions</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true"></a>SPN_P &lt;-<span class="st"> </span><span class="kw">spn_P_epiSIS_network</span>(<span class="dt">node_list =</span> node_list, <span class="dt">params =</span> theta, <span class="dt">cube =</span> cube)</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true"></a>SPN_T &lt;-<span class="st"> </span><span class="kw">spn_T_epiSIS_network</span>(<span class="dt">node_list =</span> node_list, <span class="dt">spn_P =</span> SPN_P, <span class="dt">params =</span> theta,</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true"></a>                              <span class="dt">cube =</span> cube, <span class="dt">h_move =</span> h_move, <span class="dt">m_move =</span> m_move)</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true"></a></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true"></a><span class="co"># Stoichiometry matrix</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true"></a>S &lt;-<span class="st"> </span><span class="kw">spn_S</span>(<span class="dt">spn_P =</span> SPN_P, <span class="dt">spn_T =</span> SPN_T)</span></code></pre></div>
<p>The <code>spn_P_*</code> functions (<code>spn_P_epiSIS_network()</code> used here) all return length 2 lists.The first index (<code>ix</code>) is a list equal to the number of nodes (so here, <code>length(ix) =</code> 3), which returns the uniquely numbered “places” for every life/infection/Erlang stage for the entire network. These correspond to every location in the state vector used during the simulation.</p>
<p>The second item in those lists (<code>u</code>) is the name for every “place”. It is a combination of genotype and node number for eggs, larvae, pupae, unmated females, and male stages. For females, it is a combination of their genotype, their mate genotype, their infection status, and the node number. For humans, it is just infection status and node number.</p>
<p>The <code>spn_T_*</code> functions (<code>spn_T_epiSIS_network()</code> used here) returns a length 2 list. The first element (<code>T</code>) is a list of parameters for every possible state transition. These are not the rates of transition, just the fact that it is possible and parameters necessary to describe them. The second element (<code>v</code>) is the name for each of those transitions.</p>
</div>
<div id="equilibria_haz" class="section level2">
<h2>Equilibrium Conditions and Hazard Functions</h2>
<p>Remember, these are node-by-node equilibria, not an equilibrium over the entire network. We expect some burn-in at the beginning of the simulation as the network reaches equilibrium.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a><span class="co"># SEI mosquitoes and SIS humans equilibrium</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true"></a><span class="co">#  outputs required parameters in the named list &quot;params&quot;</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true"></a><span class="co">#  outputs intial equilibrium for adv users, &quot;init</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true"></a><span class="co">#  outputs properly filled initial markings, &quot;M0&quot;</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true"></a>initialCons &lt;-<span class="st"> </span><span class="kw">equilibrium_SEI_SIS</span>(<span class="dt">params =</span> theta, <span class="dt">node_list =</span> node_list,</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true"></a>                                   <span class="dt">NF =</span> <span class="dv">500</span>, <span class="dt">phi =</span> <span class="fl">0.5</span>, <span class="dt">NH =</span> <span class="dv">1000</span>, <span class="dt">pop_ratio_H =</span> <span class="fl">0.15</span>,</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true"></a>                                   <span class="dt">log_dd =</span> <span class="ot">TRUE</span>, <span class="dt">spn_P =</span> SPN_P, <span class="dt">cube =</span> cube)</span></code></pre></div>
<p>The <code>equilibrium_*</code> functions (<code>equilibrium_SEI_SIS()</code> used here) return a length 3 list. The first element (<code>params</code>) are all of the parameters required for the simulation. These include user-supplied parameters from before (<code>theta</code>), as well as derived parameters, such as the density-dependent parameter. The second element is a matrix of equilibrium populations. The rows are nodes, the columns are life stages. Any stage that does not exist in a node (human stages in mosquito-only nodes, mosquito stages in human-only nodes) have <code>NA</code> values. The final element is <code>M0</code>, the initial marking for the Petri Net. This takes all of the equilibrium values from the second element, and using the places indices, distributes populations to their appropriate location in the Petri Net framework.</p>
<p>Next, we setup the movement rates for mosquitoes and humans, to their respective nodes. See the <a href="epi-network.html">epi-network</a> vignette for a detailed explanation.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a><span class="co"># calculate movement rates and movement probabilities</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a>gam &lt;-<span class="st"> </span><span class="kw">calc_move_rate</span>(<span class="dt">mu =</span> initialCons<span class="op">$</span>params<span class="op">$</span>muF, <span class="dt">P =</span> <span class="fl">0.05</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true"></a><span class="co"># set mosquito movement rates/probabilities</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true"></a><span class="co">#  mosquitoes exist in nodes 1 and 2, not 3</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true"></a>mr_mosy &lt;-<span class="st"> </span><span class="kw">c</span>(gam, gam, <span class="ot">NaN</span>)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true"></a>mp_mosy &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">sparseMatrix</span>(<span class="dt">i =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">j =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">1</span>), <span class="dt">x =</span> <span class="dv">1</span>, <span class="dt">dims =</span> <span class="kw">dim</span>(m_move))</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true"></a><span class="co"># set human movement rates/probabilities</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true"></a><span class="co">#  humans exist in nodes 2 and 3, not 1</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true"></a>mr_human &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="ot">NaN</span>, <span class="dv">1</span><span class="op">/</span><span class="dv">7</span>, <span class="dv">1</span><span class="op">/</span><span class="dv">7</span>)</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true"></a>mp_human &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">sparseMatrix</span>(<span class="dt">i =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>), <span class="dt">j =</span> <span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">2</span>), <span class="dt">x =</span> <span class="dv">1</span>, <span class="dt">dims =</span> <span class="kw">dim</span>(h_move))</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true"></a></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true"></a><span class="co"># put rates and probs into the parameter list</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true"></a>initialCons<span class="op">$</span>params<span class="op">$</span>mosquito_move_rates &lt;-<span class="st"> </span>mr_mosy</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true"></a>initialCons<span class="op">$</span>params<span class="op">$</span>mosquito_move_probs &lt;-<span class="st"> </span>mp_mosy</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true"></a>initialCons<span class="op">$</span>params<span class="op">$</span>human_move_rates &lt;-<span class="st"> </span>mr_human</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true"></a>initialCons<span class="op">$</span>params<span class="op">$</span>human_move_probs &lt;-<span class="st"> </span>mp_human</span></code></pre></div>
<p>Now that all the necessary parameters have been added to the named list <code>initialCons$params</code>, we generate the hazard functions, using the function <code>spn_hazards()</code>. By specifying <code>log_dd = TRUE</code>, we use logistic density dependence for these simulations.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a><span class="co"># exact hazards for integer-valued state space</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true"></a>exact_hazards &lt;-<span class="st"> </span><span class="kw">spn_hazards</span>(<span class="dt">spn_P =</span> SPN_P, <span class="dt">spn_T =</span> SPN_T, <span class="dt">cube =</span> cube,</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true"></a>                             <span class="dt">params =</span> initialCons<span class="op">$</span>params, <span class="dt">type =</span> <span class="st">&quot;SIS&quot;</span>,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true"></a>                             <span class="dt">log_dd =</span> <span class="ot">TRUE</span>, <span class="dt">exact =</span> <span class="ot">TRUE</span>, <span class="dt">tol =</span> <span class="ot">NaN</span>,</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true"></a>                             <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>The hazard function (<code>spn_hazards()</code>) returns a list of functions that define the rates of every transition. It is the same length as the first element of the <code>SPN_T</code>, i.e., one rate for every possible transition in the Petri Net.</p>
</div>
<div id="sim" class="section level2">
<h2>Simulation of Fully Specified SPN Model</h2>
<p>Before running any simulations, we need to create some releases, and then setup the folder structure.</p>
<p>We will release 50 adult females with homozygous recessive alleles 2 times, every 7 days, starting at day 10, in node 1 (this allows us to see movement from node 1 to node 2, before we see an impact on human disease transmission). Remember, it is critically important that <strong>the event names match a place name</strong> in the simulation. The simulation function checks this and will throw an error if the event name does not exist as a place in the simulation. This format is used in <strong>MGDrivE2</strong> for consistency with solvers in <code>deSolve</code>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true"></a><span class="co"># releases</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true"></a>r_times &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">10</span>, <span class="dt">length.out =</span> <span class="dv">2</span>, <span class="dt">by =</span> <span class="dv">7</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true"></a>r_size &lt;-<span class="st"> </span><span class="dv">50</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true"></a>events &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="st">&quot;var&quot;</span> =<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;F_&quot;</span>, cube<span class="op">$</span>releaseType, <span class="st">&quot;_&quot;</span>, cube<span class="op">$</span>wildType, <span class="st">&quot;_S_1&quot;</span>),</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true"></a>                     <span class="st">&quot;time&quot;</span> =<span class="st"> </span>r_times,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true"></a>                     <span class="st">&quot;value&quot;</span> =<span class="st"> </span>r_size,</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true"></a>                     <span class="st">&quot;method&quot;</span> =<span class="st"> &quot;add&quot;</span>,</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true"></a>                     <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>Next, we need to setup the file structure for CSV simulations.</p>
<p>We <strong>recommend</strong> setting up a new folder for any simulations. This ensures a clean location, and in the chance someone deletes everything in the folder, there is nothing important there. Inside the main folder (<code>main_out</code>), we need three folders, for each stage of the analysis. We generally call these <code>raw</code>, <code>traces</code>, and <code>analyzed</code>, to describe their purpose. The <code>raw</code> folder holds all of the raw output from the simulation. It is further divided into repetition folders. These can be named anything, we use a numeric naming scheme that sorts properly in file systems. The <code>traces</code> folder holds the first set analysis output. It starts empty, the provided functions setup the necessary folders, and each repetition will be split into the appropriate life-stage and node designation. The final folder, <code>analyzed</code>, holds the summary statistics from our simulations.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true"></a><span class="co"># main output folder</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true"></a>main_out &lt;-<span class="st"> </span><span class="kw">tempdir</span>()</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true"></a><span class="co"># folders for each stage of analysis</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true"></a>analysis_out &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;raw&quot;</span>, <span class="st">&quot;traces&quot;</span>, <span class="st">&quot;analyzed&quot;</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true"></a><span class="co"># repetitions, 3 of them</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true"></a>rep_out &lt;-<span class="st"> </span><span class="kw">formatC</span>(<span class="dt">x =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dt">width=</span><span class="dv">3</span>, <span class="dt">format=</span><span class="st">&#39;d&#39;</span>, <span class="dt">flag=</span><span class="st">&#39;0&#39;</span>)</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true"></a></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true"></a><span class="co"># build analysis folders</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true"></a>analysis_folders &lt;-<span class="st"> </span><span class="kw">file.path</span>(main_out, analysis_out)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true"></a><span class="cf">for</span>(i <span class="cf">in</span> analysis_folders){ <span class="kw">dir.create</span>(<span class="dt">path =</span> i, <span class="dt">recursive =</span> <span class="ot">TRUE</span>) }</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true"></a></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true"></a><span class="co"># build repetition folders</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true"></a>rep_folders &lt;-<span class="st"> </span><span class="kw">file.path</span>(analysis_folders[<span class="dv">1</span>], rep_out)</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true"></a><span class="cf">for</span>(i <span class="cf">in</span> rep_folders){ <span class="kw">dir.create</span>(i) }</span></code></pre></div>
<div id="soln_tau" class="section level3">
<h3>Tau-leaping Simulation</h3>
<p>We start by running the simulations. As explained in the <a href="inhomogeneous.html">inhomogeneous</a> vignette, there are two ways to perform repetitions - using the internal repetition wrapper, or using a parallel loop. By providing output folders to <code>sim_trajectory_CSV()</code>, we can make use of both methods simultaneously. However, for this example, we only use the internal repetitions wrapper.</p>
<p>Additionally, we need to specify what life-stages to store. For this example, we will output all of the stages. However, we recommend only saving the stages necessary for the experiment at hand. Each stage to print has a one-letter designation for the function.</p>
<table>
<thead>
<tr class="header">
<th align="center">One-Letter Designation</th>
<th align="left">Stage Name</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><code>E</code></td>
<td align="left">Egg stage</td>
</tr>
<tr class="even">
<td align="center"><code>L</code></td>
<td align="left">Larval stage</td>
</tr>
<tr class="odd">
<td align="center"><code>P</code></td>
<td align="left">Pupal stage</td>
</tr>
<tr class="even">
<td align="center"><code>M</code></td>
<td align="left">Adult male stage</td>
</tr>
<tr class="odd">
<td align="center"><code>U</code></td>
<td align="left">Adult, unmated female stage</td>
</tr>
<tr class="even">
<td align="center"><code>F</code></td>
<td align="left">Adult, mated female stage</td>
</tr>
<tr class="odd">
<td align="center"><code>H</code></td>
<td align="left">Adult humans</td>
</tr>
</tbody>
</table>
<p>The mated female stage (<code>F</code>) will print infection states to different files. This keeps individual files from becoming too unwieldy.</p>
<p>Finally, we run three stochastic realizations of the simulation, using the <code>tau</code> sampler with <span class="math inline">\(\Delta t = 0.1\)</span>, approximating 10 jumps per day.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a><span class="co"># delta t</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true"></a>dt_stoch &lt;-<span class="st"> </span><span class="fl">0.1</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true"></a><span class="co"># run tau-leaping simulation</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true"></a><span class="kw">sim_trajectory_CSV</span>(<span class="dt">x0 =</span> initialCons<span class="op">$</span>M0, <span class="dt">tmax =</span> tmax, <span class="dt">dt =</span> dt,</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true"></a>                   <span class="dt">dt_stoch =</span> dt_stoch, <span class="dt">folders =</span> rep_folders,</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true"></a>                   <span class="dt">stage =</span> <span class="kw">c</span>(<span class="st">&quot;E&quot;</span>, <span class="st">&quot;L&quot;</span>, <span class="st">&quot;P&quot;</span>, <span class="st">&quot;M&quot;</span>, <span class="st">&quot;U&quot;</span>, <span class="st">&quot;F&quot;</span>, <span class="st">&quot;H&quot;</span>), <span class="dt">S =</span> S,</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true"></a>                   <span class="dt">hazards =</span> exact_hazards, <span class="dt">events =</span> events, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true"></a><span class="co">#&gt; Warning in base_events(x0 = x0, events = events, dt = dt): event times do not correspond to a multiple of dt.</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true"></a><span class="co">#&gt; event times will be rounded up to the nearest time-step!</span></span></code></pre></div>
<p>We can see all of the files returned by the simulation.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true"></a><span class="co"># list all files from the first repetition</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true"></a><span class="kw">list.files</span>(<span class="dt">path =</span> rep_folders[<span class="dv">1</span>], <span class="dt">full.names =</span> <span class="ot">FALSE</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true"></a><span class="co">#&gt; [1] &quot;E.csv&quot;  &quot;FE.csv&quot; &quot;FI.csv&quot; &quot;FS.csv&quot; &quot;H.csv&quot;  &quot;L.csv&quot;  &quot;M.csv&quot;  &quot;P.csv&quot; </span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true"></a><span class="co">#&gt; [9] &quot;U.csv&quot;</span></span></code></pre></div>
<p>Each file begins with the one-letter signifier. Notice that the female stage (<code>F</code>) has been expanded into three stages, one for each infection (<strong>SEI</strong>) stage. These names must stay the same, as the following analysis only matches the outputs from <code>sim_trajectory_CSV()</code>.</p>
<p>To see how the files are organized, we will open two of them: the egg stage (<code>E.csv</code>) and the susceptible female stage (<code>FS.csv</code>).</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true"></a><span class="co"># read the eggs from repetition 1</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true"></a>egg_stage &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="kw">file.path</span>(rep_folders[<span class="dv">1</span>], <span class="st">&quot;E.csv&quot;</span>), <span class="dt">header =</span> <span class="ot">TRUE</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true"></a><span class="co"># read susceptible females from repetition 1</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true"></a>fs_stage &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="kw">file.path</span>(rep_folders[<span class="dv">1</span>], <span class="st">&quot;FS.csv&quot;</span>), <span class="dt">header =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>First, we will look at the dimensions of each file. Everything is printed at the same rate, <code>dt</code>, so the length of the files should match, but the number of columns is significantly different.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true"></a><span class="co"># eggs</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true"></a><span class="kw">dim</span>(egg_stage)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true"></a><span class="co">#&gt; [1] 26 13</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true"></a><span class="co"># susceptible females</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true"></a><span class="kw">dim</span>(fs_stage)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true"></a><span class="co">#&gt; [1] 26 19</span></span></code></pre></div>
<p>Every file begins with a “Time” column, followed by the stage designations. It is these stage designations that explain why female files are so much larger than the rest. As described earlier, the names for each place are a combination of genotypes and node for most stages, but for adult females, places are labeled by 3 designations. The column labels in the egg file are:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true"></a><span class="kw">colnames</span>(egg_stage)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true"></a><span class="co">#&gt;  [1] &quot;Time&quot;    &quot;E1_AA_1&quot; &quot;E2_AA_1&quot; &quot;E1_Aa_1&quot; &quot;E2_Aa_1&quot; &quot;E1_aa_1&quot; &quot;E2_aa_1&quot;</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true"></a><span class="co">#&gt;  [8] &quot;E1_AA_2&quot; &quot;E2_AA_2&quot; &quot;E1_Aa_2&quot; &quot;E2_Aa_2&quot; &quot;E1_aa_2&quot; &quot;E2_aa_2&quot;</span></span></code></pre></div>
<p>First column is “Time”, and each subsequent column is the Erlang-stage, genotype, and node. These labels correspond to the labels in the <code>SPN_P$u</code> list, and the labels on the state vector, <code>M0</code>.</p>
<p>Looking at the females output, we see how this state space gets large quickly.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true"></a><span class="kw">colnames</span>(fs_stage)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true"></a><span class="co">#&gt;  [1] &quot;Time&quot;        &quot;F_AA_AA_S_1&quot; &quot;F_AA_Aa_S_1&quot; &quot;F_AA_aa_S_1&quot; &quot;F_Aa_AA_S_1&quot;</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true"></a><span class="co">#&gt;  [6] &quot;F_Aa_Aa_S_1&quot; &quot;F_Aa_aa_S_1&quot; &quot;F_aa_AA_S_1&quot; &quot;F_aa_Aa_S_1&quot; &quot;F_aa_aa_S_1&quot;</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true"></a><span class="co">#&gt; [11] &quot;F_AA_AA_S_2&quot; &quot;F_AA_Aa_S_2&quot; &quot;F_AA_aa_S_2&quot; &quot;F_Aa_AA_S_2&quot; &quot;F_Aa_Aa_S_2&quot;</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true"></a><span class="co">#&gt; [16] &quot;F_Aa_aa_S_2&quot; &quot;F_aa_AA_S_2&quot; &quot;F_aa_Aa_S_2&quot; &quot;F_aa_aa_S_2&quot;</span></span></code></pre></div>
<p>Again, the “Time” column is first, followed by female genotypes, mate genotypes, infection stage, and node location. This is why female files have been split into different infection stages.</p>
</div>
<div id="split_agg" class="section level3">
<h3>Trace Analysis</h3>
<p>The raw analysis is large, and difficult to work with. So, as a first reduction, <strong>MGDrivE2</strong> provides the function <code>split_aggregate_CSV()</code> to organize the raw output. This function takes the <code>raw</code> folder, duplicates the folder structure in the <code>traces</code> folder, and reduces the raw output by several metrics. The aquatic stages are summarized by genotype, with the option to summarize them by Erlang-stage if desired, and split into patches. Adult female stages are summarized by female genotype, summing over male mates, and then split by patch as well. This function expects file names output from the previous function, and by default looks for all of the output. If anything is not found, it is skipped. Additionally, the user can specify subsets of the <code>raw</code> output and run different analyses over each.</p>
<p>We designate the folder <code>traces</code> because we can plot population traces from the output. These are the individual dynamics of each genotype and each stage for every simulation.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true"></a><span class="co"># split everything by patch, aggregate by genotype</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true"></a><span class="kw">split_aggregate_CSV</span>(<span class="dt">read_dir =</span> analysis_folders[<span class="dv">1</span>], <span class="dt">write_dir =</span> analysis_folders[<span class="dv">2</span>],</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true"></a>                    <span class="dt">spn_P =</span> SPN_P, <span class="dt">tmax =</span> tmax, <span class="dt">dt =</span> dt, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>Looking at the <code>traces</code> folder, we see the folder structure is identical to the <code>raw</code> folder.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true"></a><span class="co"># don&#39;t list parent directory</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true"></a><span class="kw">list.dirs</span>(<span class="dt">path =</span> analysis_folders[<span class="dv">2</span>], <span class="dt">recursive =</span> <span class="ot">FALSE</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true"></a><span class="co">#&gt; [1] &quot;/var/folders/qv/y1wxv5g94gsgkt5dm8c08vy00000gn/T//RtmpUAF1Jc/traces/001&quot;</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true"></a><span class="co">#&gt; [2] &quot;/var/folders/qv/y1wxv5g94gsgkt5dm8c08vy00000gn/T//RtmpUAF1Jc/traces/002&quot;</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true"></a><span class="co">#&gt; [3] &quot;/var/folders/qv/y1wxv5g94gsgkt5dm8c08vy00000gn/T//RtmpUAF1Jc/traces/003&quot;</span></span></code></pre></div>
<p>We can look at the output in the first repetition folder.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true"></a><span class="kw">list.files</span>(<span class="dt">path =</span> <span class="kw">file.path</span>(analysis_folders[<span class="dv">2</span>], rep_out[<span class="dv">1</span>]))</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true"></a><span class="co">#&gt;  [1] &quot;E_0001.csv&quot;  &quot;E_0002.csv&quot;  &quot;FE_0001.csv&quot; &quot;FE_0002.csv&quot; &quot;FI_0001.csv&quot;</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true"></a><span class="co">#&gt;  [6] &quot;FI_0002.csv&quot; &quot;FS_0001.csv&quot; &quot;FS_0002.csv&quot; &quot;H_0002.csv&quot;  &quot;H_0003.csv&quot; </span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true"></a><span class="co">#&gt; [11] &quot;L_0001.csv&quot;  &quot;L_0002.csv&quot;  &quot;M_0001.csv&quot;  &quot;M_0002.csv&quot;  &quot;P_0001.csv&quot; </span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true"></a><span class="co">#&gt; [16] &quot;P_0002.csv&quot;  &quot;U_0001.csv&quot;  &quot;U_0002.csv&quot;</span></span></code></pre></div>
<p>Notice, every file begins with our one/two letter stage designation. The number following it is the node number. Remember, our network has 3 nodes, m, b, h. We expect that mosquito stages are found in nodes one and two, while humans are found in nodes 2 and 3.</p>
<p>We look at the reduced versions of the same two files we opened previously to compare the differences.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true"></a><span class="co"># read the eggs from repetition 1</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true"></a>egg_stage &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="kw">file.path</span>(analysis_folders[<span class="dv">2</span>], rep_out[<span class="dv">1</span>], <span class="st">&quot;E_0001.csv&quot;</span>),</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true"></a>                      <span class="dt">header =</span> <span class="ot">TRUE</span>)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true"></a><span class="co"># read susceptible females from repetition 1</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true"></a>fs_stage &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="kw">file.path</span>(analysis_folders[<span class="dv">2</span>], rep_out[<span class="dv">1</span>], <span class="st">&quot;FS_0001.csv&quot;</span>),</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true"></a>                     <span class="dt">header =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>The length of each file is the same, since we need every time point, but the number of columns is significantly less.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true"></a><span class="co"># eggs</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true"></a><span class="kw">dim</span>(egg_stage)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true"></a><span class="co">#&gt; [1] 26  4</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true"></a></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true"></a><span class="co"># susceptible females</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true"></a><span class="kw">dim</span>(fs_stage)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true"></a><span class="co">#&gt; [1] 26  4</span></span></code></pre></div>
<p>Again, every file begins with a “Time” column, but now all files have the same remaining columns as well, just the genotypes.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true"></a><span class="co"># eggs</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true"></a><span class="kw">colnames</span>(egg_stage)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true"></a><span class="co">#&gt; [1] &quot;Time&quot; &quot;AA&quot;   &quot;Aa&quot;   &quot;aa&quot;</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true"></a><span class="co"># females</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true"></a><span class="kw">colnames</span>(fs_stage)</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true"></a><span class="co">#&gt; [1] &quot;Time&quot; &quot;AA&quot;   &quot;Aa&quot;   &quot;aa&quot;</span></span></code></pre></div>
<p>Since our genotypes stay the same throughout the population, every mosquito file now has the same columns names, “Time” and then genotypes. Humans are still separated by infection stage, since that is the only way we keep track of them.</p>
</div>
<div id="summary" class="section level3">
<h3>Summary Analysis</h3>
<p>Traces make nice squiggly plots, but are not particularly helpful otherwise. Our final step is calculating summary statistics from the simulations. <strong>MGDrivE2</strong> provides a function (<code>summarize_stats_CSV()</code>) to calculate the mean and quantiles over all simulations. It reads the output of <code>split_aggregate_CSV()</code>, and returns the desired summaries.</p>
<p>Quantiles are calculated empirically, using algorithm 8 from the built-in quantile function.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true"></a><span class="co"># mean and 95% quantiles</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true"></a><span class="kw">summarize_stats_CSV</span>(<span class="dt">read_dir =</span> analysis_folders[<span class="dv">2</span>], <span class="dt">write_dir =</span> analysis_folders[<span class="dv">3</span>],</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true"></a>                    <span class="dt">spn_P =</span> SPN_P, <span class="dt">tmax =</span> tmax, <span class="dt">dt =</span> dt, <span class="dt">mean =</span> <span class="ot">TRUE</span>,</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true"></a>                    <span class="dt">quantiles =</span> <span class="kw">c</span>(<span class="fl">0.025</span>, <span class="fl">0.975</span>), <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>Since repetitions are summarized, we do not have the same directory structure in the <code>analyzed</code> folder, just the summaries.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true"></a><span class="kw">list.files</span>(<span class="dt">path =</span> analysis_folders[<span class="dv">3</span>])</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true"></a><span class="co">#&gt;  [1] &quot;E_Mean_0001.csv&quot;            &quot;E_Mean_0002.csv&quot;           </span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true"></a><span class="co">#&gt;  [3] &quot;E_Quantile_00250_0001.csv&quot;  &quot;E_Quantile_00250_0002.csv&quot; </span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true"></a><span class="co">#&gt;  [5] &quot;E_Quantile_09750_0001.csv&quot;  &quot;E_Quantile_09750_0002.csv&quot; </span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true"></a><span class="co">#&gt;  [7] &quot;FE_Mean_0001.csv&quot;           &quot;FE_Mean_0002.csv&quot;          </span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true"></a><span class="co">#&gt;  [9] &quot;FE_Quantile_00250_0001.csv&quot; &quot;FE_Quantile_00250_0002.csv&quot;</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true"></a><span class="co">#&gt; [11] &quot;FE_Quantile_09750_0001.csv&quot; &quot;FE_Quantile_09750_0002.csv&quot;</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true"></a><span class="co">#&gt; [13] &quot;FI_Mean_0001.csv&quot;           &quot;FI_Mean_0002.csv&quot;          </span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true"></a><span class="co">#&gt; [15] &quot;FI_Quantile_00250_0001.csv&quot; &quot;FI_Quantile_00250_0002.csv&quot;</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true"></a><span class="co">#&gt; [17] &quot;FI_Quantile_09750_0001.csv&quot; &quot;FI_Quantile_09750_0002.csv&quot;</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true"></a><span class="co">#&gt; [19] &quot;FS_Mean_0001.csv&quot;           &quot;FS_Mean_0002.csv&quot;          </span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true"></a><span class="co">#&gt; [21] &quot;FS_Quantile_00250_0001.csv&quot; &quot;FS_Quantile_00250_0002.csv&quot;</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true"></a><span class="co">#&gt; [23] &quot;FS_Quantile_09750_0001.csv&quot; &quot;FS_Quantile_09750_0002.csv&quot;</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true"></a><span class="co">#&gt; [25] &quot;H_Mean_0002.csv&quot;            &quot;H_Mean_0003.csv&quot;           </span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true"></a><span class="co">#&gt; [27] &quot;H_Quantile_00250_0002.csv&quot;  &quot;H_Quantile_00250_0003.csv&quot; </span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true"></a><span class="co">#&gt; [29] &quot;H_Quantile_09750_0002.csv&quot;  &quot;H_Quantile_09750_0003.csv&quot; </span></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true"></a><span class="co">#&gt; [31] &quot;L_Mean_0001.csv&quot;            &quot;L_Mean_0002.csv&quot;           </span></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true"></a><span class="co">#&gt; [33] &quot;L_Quantile_00250_0001.csv&quot;  &quot;L_Quantile_00250_0002.csv&quot; </span></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true"></a><span class="co">#&gt; [35] &quot;L_Quantile_09750_0001.csv&quot;  &quot;L_Quantile_09750_0002.csv&quot; </span></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true"></a><span class="co">#&gt; [37] &quot;M_Mean_0001.csv&quot;            &quot;M_Mean_0002.csv&quot;           </span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true"></a><span class="co">#&gt; [39] &quot;M_Quantile_00250_0001.csv&quot;  &quot;M_Quantile_00250_0002.csv&quot; </span></span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true"></a><span class="co">#&gt; [41] &quot;M_Quantile_09750_0001.csv&quot;  &quot;M_Quantile_09750_0002.csv&quot; </span></span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true"></a><span class="co">#&gt; [43] &quot;P_Mean_0001.csv&quot;            &quot;P_Mean_0002.csv&quot;           </span></span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true"></a><span class="co">#&gt; [45] &quot;P_Quantile_00250_0001.csv&quot;  &quot;P_Quantile_00250_0002.csv&quot; </span></span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true"></a><span class="co">#&gt; [47] &quot;P_Quantile_09750_0001.csv&quot;  &quot;P_Quantile_09750_0002.csv&quot; </span></span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true"></a><span class="co">#&gt; [49] &quot;U_Mean_0001.csv&quot;            &quot;U_Mean_0002.csv&quot;           </span></span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true"></a><span class="co">#&gt; [51] &quot;U_Quantile_00250_0001.csv&quot;  &quot;U_Quantile_00250_0002.csv&quot; </span></span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true"></a><span class="co">#&gt; [53] &quot;U_Quantile_09750_0001.csv&quot;  &quot;U_Quantile_09750_0002.csv&quot;</span></span></code></pre></div>
<p>Again, the stage designation is at the front of every file name. Next, either <code>Mean</code> or <code>Quantile</code>. If <code>Mean</code>, then the node designation. If <code>Quantile</code>, which quantile, and then the node designation.</p>
<p>We will open the egg and susceptible females files, to see the summaries, but the structure of these files is the same as the trace files.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true"></a><span class="co"># read the eggs from repetition 1</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true"></a>egg_stage &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="kw">file.path</span>(analysis_folders[<span class="dv">3</span>], <span class="st">&quot;E_Mean_0001.csv&quot;</span>),</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true"></a>                      <span class="dt">header =</span> <span class="ot">TRUE</span>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true"></a><span class="co"># read susceptible females from repetition 1</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true"></a>fs_stage &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="kw">file.path</span>(analysis_folders[<span class="dv">3</span>], <span class="st">&quot;FS_Mean_0001.csv&quot;</span>),</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true"></a>                     <span class="dt">header =</span> <span class="ot">TRUE</span>)</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true"></a></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true"></a><span class="co"># eggs</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true"></a><span class="kw">dim</span>(egg_stage)</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true"></a><span class="co">#&gt; [1] 26  4</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true"></a></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true"></a><span class="co"># susceptible females</span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true"></a><span class="kw">dim</span>(fs_stage)</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true"></a><span class="co">#&gt; [1] 26  4</span></span></code></pre></div>
<p>The dimensions are the same as the trace files. We also see that the column names are the same as well.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true"></a><span class="co"># eggs</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true"></a><span class="kw">colnames</span>(egg_stage)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true"></a><span class="co">#&gt; [1] &quot;Time&quot; &quot;AA&quot;   &quot;Aa&quot;   &quot;aa&quot;</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true"></a></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true"></a><span class="co"># females</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true"></a><span class="kw">colnames</span>(fs_stage)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true"></a><span class="co">#&gt; [1] &quot;Time&quot; &quot;AA&quot;   &quot;Aa&quot;   &quot;aa&quot;</span></span></code></pre></div>
<p>The plotting utilities provided by <strong>MGDrivE2</strong> do not work on CSV output. Our assumption is that, if one is generating enough data to use the CSV utilities, then there are specific plotting requirements.</p>
<!-- ```{r, echo=FALSE}
# Cleanup
unlink(x = main_out, recursive = TRUE)
``` -->
</div>
<div id="ref" class="section level3">
<h3>References</h3>
<ul>
<li>Smith, D. L., &amp; McKenzie, F. E. (2004). Statics and dynamics of malaria infection in Anopheles mosquitoes. Malaria journal, 3(1), 13.</li>
</ul>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
